lib/reversetunnel/agentpool.go

```diff
@@ -135,207 +165,433 @@ func NewAgentPool(ctx context.Context, cfg AgentPoolConfig) (*AgentPool, error)
-			return
-		case <-limiter.After():
-			limiter.Inc()
+
+		err = p.waitForBackoff(p.ctx, p.events)
+		if err != nil {
+			p.log.WithError(err).Debugf("Failed to wait for backoff.")
 		}
 	}
 }
 
-func (m *AgentPool) withLock(f func()) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	f()
+// connectAgent connects a new agent and processes any agent events blocking until a
+// new agent is connected or an error occurs.
+func (p *AgentPool) connectAgent(ctx context.Context, leases <-chan track.Lease, events <-chan Agent) (*agent, error) {
+	lease, err := p.waitForLease(ctx, leases, events)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Ensure the lease is released on error.
+	defer func() {
+		if err != nil {
+			lease.Release()
+		}
+	}()
+
+	err = p.processEvents(ctx, events)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	agent, err := p.newAgent(ctx, p.tracker, lease)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	err = agent.Start(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return agent, nil
 }
 
-type matchAgentFn func(a *Agent) bool
+func (p *AgentPool) updateRuntimeConfig(ctx context.Context) {
+	netConfig, err := p.AccessPoint.GetClusterNetworkingConfig(ctx)
+	if err != nil {
+		p.log.WithError(err).Warnf("Failed to get latest cluster networking config.")
+		return
+	}
 
-func (m *AgentPool) closeAgents() {
-	m.agents = filterAndClose(m.agents, func(*Agent) bool { return true })
+	p.runtimeConfig.update(netConfig)
 }
 
-func filterAndClose(agents []*Agent, matchAgent matchAgentFn) []*Agent {
-	var filtered []*Agent
-	for i := range agents {
-		agent := agents[i]
-		if matchAgent(agent) {
-			agent.log.Debugf("Pool is closing agent.")
-			if err := agent.Close(); err != nil {
-				agent.log.WithError(err).Warnf("Failed to close agent")
-			}
-		} else {
-			filtered = append(filtered, agent)
+// processEvents handles all events in the queue. Unblocking when a new agent
+// is required.
+func (p *AgentPool) processEvents(ctx context.Context, events <-chan Agent) error {
+	// Processes any queued events without blocking.
+	for {
+		select {
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		case agent := <-events:
+			p.handleEvent(ctx, agent)
+			continue
+		default:
 		}
+		break
 	}
 
-	if len(filtered) <= 0 {
+	p.updateRuntimeConfig(ctx)
+	p.disconnectAgents()
+	if p.isAgentRequired() {
 		return nil
 	}
 
-	return filtered
+	// Continue to process new events until an agent is required.
+	for {
+		p.log.Debugf("processing events...")
+		select {
+		case <-ctx.Done():
+			return trace.Wrap(ctx.Err())
+		case agent := <-events:
+			p.handleEvent(ctx, agent)
+
+			p.updateRuntimeConfig(ctx)
+			p.disconnectAgents()
+			if p.isAgentRequired() {
+				return nil
+			}
+		case <-p.timer.C:
+			agent, ok := p.active.poplen(0)
+			if !ok {
+				continue
+			}
+			p.log.Debugf("stopping agent...")
+			go func() {
+				agent.Stop()
+				p.wg.Done()
+			}()
+			time.Sleep(time.Second * 15)
```

Comment:
is this sleep necessary?

Reply:
Definitely not. Good catch this was debug code. Fixed here [e50ccda](https://github.com/gravitational/teleport/pull/10444/commits/e50ccdaf31a0657aea21e44fbb4d7264a8825e8c)